### PR TITLE
[Large 2D load for GEMM - #3] Support the repCluster field in tt.dot operation lowering.

### PR DESCRIPTION
**Pull request by chengjunlu**
Support the repCluster field in tt.dot operation lowering.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 1]
```diff
@@ -180,16 +180,20 @@ class DotOpDPASConversionHelper {
...
 
       fc.at({m, n}) = ret;
     };
 
+    auto repCluster = dpasEncoding.getRepCluster();
     for (int k = 0; k < repK; ++k)
       for (int m = 0; m < repM; ++m)
         for (int n = 0; n < repN; ++n)
-          generateDPASOp(m, n, k);
+          for (int repRow = 0; repRow < repCluster[0]; repRow++)
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:193]: use prefix increment (consistency with other loops in this loop nest)

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 2]
```diff
@@ -226,19 +230,22 @@ class DotOpDPASConversionHelper {
...
-        auto vecType = cast<mlir::VectorType>(matVal.getType());
-        auto valTy = vecType.getElementType();
-        for (int i = 0; i < vecType.getNumElements(); ++i) {
-          auto val = extract_element(valTy, matVal, i32_val(i));
-
-          elems.push_back(val);
-        }
-      }
+      for (int k = 0; k < dim1; ++k)
+        for (int repRow = 0; repRow < repCluster[0]; repRow++)
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:237]: use prefix increment

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 3]
```diff
@@ -226,19 +230,22 @@ class DotOpDPASConversionHelper {
...
+          for (int repCol = 0; repCol < repCluster[1]; repCol++) {
+            auto matVal = vals.at(
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:239]: auto -> Value

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 4]
```diff
@@ -226,19 +230,22 @@ class DotOpDPASConversionHelper {
...
+                {m * repCluster[0] + repRow, k * repCluster[1] + repCol});
+            auto vecType = cast<mlir::VectorType>(matVal.getType());
+            auto valTy = vecType.getElementType();
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:242]: auto -> Type

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 5]
```diff
@@ -226,19 +230,22 @@ class DotOpDPASConversionHelper {
...
+            for (int i = 0; i < vecType.getNumElements(); ++i) {
+              auto val = extract_element(valTy, matVal, i32_val(i));
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:244]: auto -> Value

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 6]
```diff
@@ -248,24 +255,46 @@ class DotOpDPASConversionHelper {
...
+  ValueTable getValuesFromDotOperandLayoutStruct(Value val, int64_t outer,
+                                                 int64_t inner, Type elemTy,
+                                                 Type dotOperandType,
+                                                 uint32_t opIdx) const {
     SmallVector<Value> elems = unpackLLElements(loc, val, rewriter);
-
+    auto repCluster = dpasLayout.getRepCluster();
+    unsigned repClusterOuter;
+    unsigned repClusterInner;
+    if (opIdx == 0) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:266]: use a switch stmt

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 7]
```diff
@@ -248,24 +255,46 @@ class DotOpDPASConversionHelper {
...
+      assert(opIdx == 2);
+      repClusterOuter = repCluster[0];
+      repClusterInner = repCluster[1];
+    }
     int offset{};
     ValueTable vals;
     size_t totalElems = elems.size();
-    size_t numElemsPerOperand = totalElems / (dim0 * dim1);
+    size_t numElemsPerOperand =
+        totalElems / (outer * inner) / (repClusterOuter * repClusterInner);
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:281]: Replace last division with a multiplication.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 8]
```diff
@@ -248,24 +255,46 @@ class DotOpDPASConversionHelper {
...
     VectorType dotOpTy = vec_ty(elemTy, numElemsPerOperand);
 
-    for (int i = 0; i < dim0; ++i) {
-      for (int j = 0; j < dim1; ++j) {
-        Value matVal = rewriter.create<LLVM::UndefOp>(loc, dotOpTy);
-        for (int k = 0; k < numElemsPerOperand; ++k) {
-          matVal = insert_element(dotOpTy, matVal, elems[offset++], i32_val(k));
+    for (int i = 0; i < outer; ++i) {
+      for (int j = 0; j < inner; ++j) {
+        for (int repOuter = 0; repOuter < repClusterOuter; repOuter++) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:286]: use prefix increment for consistency

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 9]
```diff
@@ -38,11 +38,11 @@ class DotOpDPASConversionHelper {
 
     unsigned threadsPerWarp = layout.getSubGroupSize();
     unsigned opsPerChannel = layout.getOpsPerChannel();
-    SmallVector<unsigned> shapeC = layout.getShapeC();
+    SmallVector<unsigned> shapeC = layout.getDPASInstShapeC();
```
Comment (victor-eds) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:41]: I see we're copying `SmallVector` a lot in the file. Can we use `ArrayRef` instead? (TODO in a different PR, ofc)

**Review by victor-eds — COMMENTED** [review]
Can we test?

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 9 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:41]: I am not fully understand about the C++ of this part. Just used the simplest code here.
A simple question: Is it safe to refer the structure which is returned by the function as a value? (The return value is r-value?)

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 8 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 7 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 6 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 5 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 4 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 3 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 2 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 1 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: Done.

**Comment by chengjunlu** [comment]
> Can we test?

Added some basic LIT test case.

File: test/Conversion/intel/tritongpu_to_gen_dot.mlir [hunk 10]
```diff
@@ -94,3 +94,474 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 :
...
 }
+
+// -----
+
+#dpas = #triton_intel_gpu.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [1, 1], repCluster = [4, 2]}>
+#dot_operand_a = #triton_gpu.dot_op<{opIdx=0, parent=#dpas, kWidth=2}>
+#dot_operand_b = #triton_gpu.dot_op<{opIdx=1, parent=#dpas, kWidth=2}>
+
+module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 : i32, "triton_gpu.threads-per-warp" = 16 : i32} {
+  // CHECK: llvm.func spir_funccc @_Z38intel_sub_group_f16_f16_matrix_mad_k16Dv8_sDv8_iDv8_f(vector<8xi16>, vector<8xi32>, vector<8xf32>) -> vector<8xf32> attributes {passthrough = ["convergent"]}
```
Comment (etiotto) [inline test/Conversion/intel/tritongpu_to_gen_dot.mlir:105]: Very large test case. Any way to reduce it while keeping it meaningful?

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp [hunk 11]
```diff
@@ -248,24 +255,55 @@ class DotOpDPASConversionHelper {
...
-                                                 int64_t dim1, Type elemTy,
-                                                 Type dotOperandType) const {
+  ValueTable getValuesFromDotOperandLayoutStruct(Value val, int64_t outer,
+                                                 int64_t inner, Type elemTy,
+                                                 Type dotOperandType,
+                                                 uint32_t opIdx) const {
     SmallVector<Value> elems = unpackLLElements(loc, val, rewriter);
-
+    ArrayRef<unsigned> repCluster = dpasLayout.getRepCluster();
+    unsigned repClusterOuter;
```
[the remark this answers is not in the document]
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp (hunk 9 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DotOpToLLVM/DPAS.cpp:41]: we would have to change `getShapeC` and `getDPASInstShapeC` to return an `ArrayRef`. This is material for a cleanup PR.

**Pull request merged**